src/realm/sync/transform.cpp

```diff
@@ -1597,6 +1635,20 @@ DEFINE_MERGE(Instruction::Update, Instruction::Update)
             REALM_MERGE_ASSERT(left.prior_size == right.prior_size);
         }
 
+        if (left.value.type != right.value.type) {
+            // Embedded object / dictionary creation should always lose to an
+            // Update(NULL), because these structures are nested, and we need to
+            // discard any update inside the structure.
```

Comment:
This comment is kind of confusing because it talks about Update(NULL) but isnt checking it here. Are you just relying on the fact that it should never be possible for type to be Dictionary on one side and String on the other?

Reply:
Fair, I changed the phrasing a bit